### PR TITLE
Issue #518

### PR DESCRIPTION
**Pull request by ALevansSamsung**
fixes https://github.com/android-graphics/sokatoa/issues/518

https://github.com/eclipsesource/perfetto/assets/121060410/921653b8-6650-4305-81eb-a84df1ad4dba

### CODE REVIEW

**Comment by wpaul-samsung** [comment]
It's unclear why the Perfetto changes are needed as it does work like this already on ui.perfetto.dev.

**Comment by ALevansSamsung** [comment]
Because I made the changes with my initial theming work as a means to keep the overall amount of colors we had to pass to Perfetto down to a minimum.  Also, this is not how it works in perfetto, because in Perfetto they have a "sidebar color" that they use as a background color.  This solution doesn't work for us as we have different themes that we need to accommodate.  To accommodate all themes I, rather than adding a new color to the palette, flipped the text and background color of this section to still give a highlighting effect.

**Comment by wpaul-samsung** [comment]
> Because I made the changes with my initial theming work as a means to keep the overall amount of colors we had to pass to Perfetto down to a minimum. Also, this is not how it works in perfetto, because in Perfetto they have a "sidebar color" that they use as a background color. This solution doesn't work for us as we have different themes that we need to accommodate. To accommodate all themes I, rather than adding a new color to the palette, flipped the text and background color of this section to still give a highlighting effect.

OK.  It has the effect we want, so LGTM, but since I'm not familiar with any of this code I'm probably not the best person to review the changes.  I added @cdamus as a reviewer.

File: ui/src/assets/common.scss [hunk 1]
```diff
@@ -670,13 +670,13 @@ button.query-ctrl {
...
   --collapsed-transparent: fade-out(var(--main-background-color),1);
-  color: var(--main-foreground-color);
+  --expanded-background: var(--main-foreground-color);
   display: grid;
   grid-template-columns: auto 1fr;
   grid-template-rows: 1fr;
   transition: background-color 0.4s, color 0.4s;
   height: 40px;
-
+
```
Comment (cdamus) [inline ui/src/assets/common.scss:679]: ESLint squawks when a TS file has blank lines that aren't empty. Does the same happen for CSS files?

File: ui/src/frontend/track_group_panel.ts [hunk 2]
```diff
@@ -211,8 +211,7 @@ export class TrackGroupPanel extends Panel<Attrs> {
     const shell = assertExists(dom.querySelector('.shell'));
     this.shellWidth = shell.getBoundingClientRect().width;
     // TODO(andrewbb): move this to css_constants
-    this.backgroundColor =
-          getComputedStyle(dom).getPropertyValue('--collapsed-background');
+    this.backgroundColor = this.trackGroupState.collapsed ? getComputedStyle(dom).getPropertyValue('--collapsed-background'): getComputedStyle(dom).getPropertyValue('--expanded-background');
```
Comment (cdamus) [inline ui/src/frontend/track_group_panel.ts:214]: I'm sure that ESLint will complain about this line being too long. 😀

**Review by cdamus — CHANGES_REQUESTED** [review]
In light mode (which I prefer) the track groups look like dark mode when they open but light mode when they're closed.

CleanShot 2023-10-19 at 13 51 25

**Comment by ALevansSamsung** [comment]
> In light mode (which I prefer) the track groups look like dark mode when they open but light mode when they're closed.

Is that not what the issue asked for?

File: ui/src/frontend/track_group_panel.ts (hunk 2 above)
Comment (ALevansSamsung) [reply ui/src/frontend/track_group_panel.ts]: It didn't ,but I'll spread it out

**Comment by cdamus** [comment]
> > In light mode (which I prefer) the track groups look like dark mode when they open but light mode when they're closed.
> 
> Is that not what the issue asked for?

I can't imagine why. It looks terrible. But I won't argue with the requirement.

**Review by cdamus — DISMISSED** [review]

**Comment by cdamus** [comment]
Surely we don't want this. The buttons are difficult to see now on the groups.

CleanShot 2023-10-19 at 14 04 51

**Comment by wpaul-samsung** [comment]
> Surely we don't want this. The buttons are difficult to see now on the groups.
> 
> CleanShot 2023-10-19 at 14 04 51

Yeah.  That's not ideal.  I believe this requirement came in from John.  I'm not sure how important it is.  ui.perfetto.dev does do this though.  But it does not have the icons there...
image

**Comment by ALevansSamsung** [comment]
I'm working on solution for the buttons

File: ui/src/assets/common.scss [hunk 3]
```diff
@@ -456,7 +455,7 @@ $bottom-tab-padding: 10px;
       justify-content: center;
       
       &:hover:not(.disabled) {
-        background-color: var(--alt-background-color);
+        filter: brightness(2);
```
Comment (cdamus) [inline ui/src/assets/common.scss:458]: Nice. CSS colour functions FTW!

**Review by cdamus — CHANGES_REQUESTED** [review]
Whereas before we had blue buttons, now they are more black/grey, which I actually prefer over the stock Perfetto UI.

However, now the light blue roll-over background is gone (indifferent) but the roll-over effect is to make the button light grey, looking like it's _disabled_ when activated, which is confusing or at least unconventional.

(It works for the dark theme because there brightening the icon makes it look like a highlight.)

CleanShot 2023-10-19 at 16 41 07

**Review by cdamus — COMMENTED** [review]
The roll-over effect still looks to me like the button is being dimmed to indicate disablement. Do we need the roll-over at all? Am I being too picky?

CleanShot 2023-10-20 at 12 27 07

**Comment by ALevansSamsung** [comment]
The main issue is that in CSS, as far as I know, there isn't if statements.  If there was a way to get the overall luminance of a color and brighten or darken accordingly that would be best, but that doesn't seem to be possible.  The edge cases are always the issue,  for instance a pure white color + brightness would have no effect, on the other end pure black + darken would have no effect.  Even with an inverse, which I'm using in this case, a perfect middle ground grey would seem to have no effect.

**Comment by cdamus** [comment]
If computing this is not feasible (heroic effort so far) then should we just bite the bullet and push these colours down from the embedding Theia application, which supports theming? We do that already for vsync highlight colours.

**Comment by ALevansSamsung** [comment]
This issue seems to be a solution to a different issue which stemmed from annotation line rendering not rendering on EXPANDED track groups (it does on collapsed groups).  May I suggest rather than solving this in the way of changing the color of track groups based on its expanded or collapsed state, we instead focus more on making the annotation lines draw on the expanded track groups as I believe this will be a much easier solution.  

Personally I prefer the expanded and collapsed track groups being the same color as I believe the highlight in dark mode to be a little jarring.  I'm curious to see how others feel about this.

**Comment by cdamus** [comment]
> This issue seems to be a solution to a different issue which stemmed from annotation lines not rendering on EXPANDED track groups (it does on collapsed groups). Link to the associated thread on m-chat (for Samsung users) -> https://mchat.mosaic.sec.samsung.net/184060440-184061210/pl/j3hxdshd3ifyfpygpfwjfk1ejr?require_login=true

I don't see how making the expand group header black could help with rendering of annotation lines. That's apples and orange cars.

> May I suggest rather than solving this in the way of changing the color of track groups based on its expanded or collapsed state, we instead focus more on making the annotation lines draw on the expanded track groups as I believe this will be a much easier solution.

💯 agree. And together with annotation lines also the vsync columns that appear when you hit the `v` key, assuming that the appropriate source data track exists in the trace.

> Personally I prefer the expanded and collapsed track groups being the same color as I believe the highlight in dark mode to be a little jarring. I'm curious to see how others feel about this.

Again 💯 agree. The stark visual contrast changing on expansion is jarring and makes no sense to me. It looks like the UI is broken.

**Pull request closed without merging**